File: day05/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"regexp"
	"strconv"
	"strings"
)

const (
	puzzleInput = "input.txt"
)

var digitRegex = regexp.MustCompile(`\d+`)

type (
	Range struct {
		Start int
		End   int
	}

	Range2 struct {
		Dest Range
		Src  Range
	}
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatalln(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatalln(err)
		}
	}()

	var seeds []int
	var seeds2 []Range

	var rangeMaps [][]Range2

	var lastRangeMap []Range2

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if strings.HasPrefix(line, "seeds:") {
			for _, i := range digitRegex.FindAllString(line, -1) {
				num, err := strconv.Atoi(i)
				if err != nil {
					log.Fatalln(err)
				}
				seeds = append(seeds, num)
			}
			for i := 1; i < len(seeds); i += 2 {
				seeds2 = append(seeds2, Range{
					Start: seeds[i-1],
					End:   seeds[i-1] + seeds[i],
				})
			}
			continue
		} else if strings.HasSuffix(line, "map:") {
			if len(lastRangeMap) != 0 {
				rangeMaps = append(rangeMaps, lastRangeMap)
				lastRangeMap = nil
			}
			continue
		} else if line == "" {
			continue
		}
		nums := digitRegex.FindAllString(line, -1)
		if len(nums) != 3 {
			log.Fatalln("Invalid range", line)
		}
		num1, err := strconv.Atoi(nums[0])
		if err != nil {
			log.Fatalln(err)
		}
		num2, err := strconv.Atoi(nums[1])
		if err != nil {
			log.Fatalln(err)
		}
		num3, err := strconv.Atoi(nums[2])
		if err != nil {
			log.Fatalln(err)
		}
		lastRangeMap = append(lastRangeMap, Range2{
			Dest: Range{
				Start: num1,
				End:   num1 + num3,
			},
			Src: Range{
				Start: num2,
				End:   num2 + num3,
			},
		})
	}

	if err := scanner.Err(); err != nil {
		log.Fatalln(err)
	}

	if len(lastRangeMap) != 0 {
		rangeMaps = append(rangeMaps, lastRangeMap)
		lastRangeMap = nil
	}

	for _, i := range rangeMaps {
		seeds = runRange(seeds, i)
		seeds2 = runRange2(seeds2, i)
	}

	minSeed := seeds[0]
	for _, i := range seeds {
		if i < minSeed {
			minSeed = i
		}
	}
	fmt.Println("Part 1:", minSeed)

	minSeed2 := seeds2[0].Start
	for _, i := range seeds2 {
		if i.Start < minSeed2 {
			minSeed2 = i.Start
		}
	}
	fmt.Println("Part 2:", minSeed2)
}

func runRange(seeds []int, rangeMap []Range2) []int {
	res := make([]int, 0, len(seeds))
	for _, i := range seeds {
		k := i
		for _, j := range rangeMap {
			if i >= j.Src.Start && i < j.Src.End {
				k = i - j.Src.Start + j.Dest.Start
				break
			}
		}
		res = append(res, k)
	}
	return res
}

func runRange2(seeds []Range, rangeMap []Range2) []Range {
	res := make([]Range, 0, len(seeds))
	otherRanges := make([]Range, 0, len(seeds))
	for _, i := range seeds {
		nextStart := i.Start
		nextEnd := i.End
		for _, j := range rangeMap {
			if i.Start >= j.Src.Start && i.Start < j.Src.End {
				nextStart = i.Start - j.Src.Start + j.Dest.Start
				if i.End <= j.Src.End {
					nextEnd = i.End - j.Src.Start + j.Dest.Start
				} else {
					nextEnd = j.Dest.End
					otherRanges = append(otherRanges, Range{
						Start: j.Src.End,
						End:   i.End,
					})
				}
				break
			}
			if i.End > j.Src.Start && i.End <= j.Src.End {
				nextEnd = i.End - j.Src.Start + j.Dest.Start
				nextStart = j.Dest.Start
				otherRanges = append(otherRanges, Range{
					Start: i.Start,
					End:   j.Src.Start,
				})
				break
			}
			if i.Start < j.Src.Start && i.End > j.Src.End {
				nextStart = j.Dest.Start
				nextEnd = j.Dest.End
				otherRanges = append(otherRanges,
					Range{
						Start: i.Start,
						End:   j.Src.Start,
					},
					Range{
						Start: j.Src.End,
						End:   i.End,
					},
				)
				break
			}
		}
		res = append(res, Range{
			Start: nextStart,
			End:   nextEnd,
		})
	}
	if len(otherRanges) != 0 {
		res = append(res, runRange2(otherRanges, rangeMap)...)
	}
	return res
}

File: day05/src/main.rs
use once_cell::sync::Lazy;
use regex::Regex;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

static DIGIT_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r"\d+").unwrap());

struct Range {
    start: i64,
    end: i64,
}

struct Range2 {
    dest: Range,
    src: Range,
}

impl Range2 {
    fn start_in_src(&self, n: i64) -> bool {
        n >= self.src.start && n < self.src.end
    }

    fn end_in_src(&self, n: i64) -> bool {
        n > self.src.start && n <= self.src.end
    }

    fn to_dest(&self, n: i64) -> i64 {
        n - self.src.start + self.dest.start
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut seeds = Vec::new();
    let mut seeds2 = Vec::new();
    let mut range_maps = Vec::new();
    let mut last_range_map = Vec::new();

    for line in reader.lines() {
        let line = line?;
        if line.starts_with("seeds:") {
            seeds = DIGIT_REGEX
                .find_iter(&line)
                .map(|v| v.as_str().parse::<i64>())
                .collect::<Result<Vec<_>, _>>()?;
            seeds2 = seeds
                .chunks(2)
                .map(|v| match v {
                    &[a, b] => Ok(Range {
                        start: a,
                        end: a + b,
                    }),
                    _ => Err("odd number of seeds"),
                })
                .collect::<Result<Vec<_>, _>>()?;
            continue;
        } else if line.ends_with("map:") {
            if last_range_map.len() != 0 {
                range_maps.push(last_range_map);
                last_range_map = Vec::new();
            }
            continue;
        } else if line == "" {
            continue;
        }
        last_range_map.push(
            if let [a, b, c] = DIGIT_REGEX
                .find_iter(&line)
                .map(|v| v.as_str().parse::<i64>())
                .collect::<Result<Vec<_>, _>>()?[..]
            {
                Range2 {
                    dest: Range {
                        start: a,
                        end: a + c,
                    },
                    src: Range {
                        start: b,
                        end: b + c,
                    },
                }
            } else {
                Err("Invalid range")?
            },
        );
    }
    if last_range_map.len() != 0 {
        range_maps.push(last_range_map);
    }

    println!(
        "Part 1: {}",
        range_maps
            .iter()
            .fold(seeds, |acc, v| run_range(&acc, v))
            .iter()
            .min()
            .ok_or("No seeds")?
    );

    println!(
        "Part 2: {}",
        range_maps
            .iter()
            .fold(seeds2, |acc, v| run_range2(&acc, v))
            .iter()
            .min_by_key(|v| v.start)
            .ok_or("No seeds")?
            .start
    );

    Ok(())
}

fn run_range(seeds: &[i64], range_maps: &[Range2]) -> Vec<i64> {
    let mut res = Vec::with_capacity(seeds.len());
    for &i in seeds {
        let mut next = i;
        for j in range_maps {
            if j.start_in_src(i) {
                next = j.to_dest(i);
                break;
            }
        }
        res.push(next);
    }
    res
}

fn run_range2(seeds: &[Range], range_maps: &[Range2]) -> Vec<Range> {
    let mut res = Vec::with_capacity(seeds.len());
    let mut other_ranges = Vec::with_capacity(seeds.len());
    for i in seeds {
        let mut next_start = i.start;
        let mut next_end = i.end;
        for j in range_maps {
            if j.start_in_src(i.start) {
                next_start = j.to_dest(i.start);
                next_end = if j.end_in_src(i.end) {
                    j.to_dest(i.end)
                } else {
                    other_ranges.push(Range {
                        start: j.src.end,
                        end: i.end,
                    });
                    j.dest.end
                };
                break;
            }
            if j.end_in_src(i.end) {
                next_start = j.dest.start;
                next_end = j.to_dest(i.end);
                other_ranges.push(Range {
                    start: i.start,
                    end: j.src.start,
                });
                break;
            }
            if i.end > j.src.end && i.start < j.src.start {
                next_start = j.dest.start;
                next_end = j.dest.end;
                other_ranges.push(Range {
                    start: i.start,
                    end: j.src.start,
                });
                other_ranges.push(Range {
                    start: j.src.end,
                    end: i.end,
                });
                break;
            }
        }
        res.push(Range {
            start: next_start,
            end: next_end,
        });
    }
    if other_ranges.len() != 0 {
        res.append(&mut run_range2(&other_ranges, range_maps));
    }
    res
}
